File: substrate/fullscreen_open_collective_multiselect_floating_dr_compose.py
"""Fullscreen-open over collective multiselect floating DR draft-before-merge (pure).

live_dispatched / pack_dispatched / merge_executed always False.
live_execution_authorized / charge_executed / analysis_written always False.
production_router_verdict always REJECT.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.collective_multiselect_floating_dr_draft_before_merge_compose import (
    CollectiveMultiselectFloatingDrDraftBeforeMergeCompose,
    CollectiveMultiselectFloatingDrDraftBeforeMergeComposeError,
    compose_collective_multiselect_floating_dr_draft_before_merge,
)
from substrate.floating_fullscreen_open_compose import (
    FloatingFullscreenOpenCompose,
    FloatingFullscreenOpenComposeError,
    compose_floating_fullscreen_open,
)


class FullscreenOpenCollectiveMultiselectFloatingDrComposeError(ValueError):
    """Fail-closed validation for fullscreen + collective multiselect pack."""


@dataclass(frozen=True)
class FullscreenOpenCollectiveMultiselectFloatingDrCompose:
    session_id: str
    parent_asset_id: str
    week_id: str
    asset_id: str
    title: str
    account_id: str
    fullscreen: FloatingFullscreenOpenCompose
    collective_pack: CollectiveMultiselectFloatingDrDraftBeforeMergeCompose
    session_aligned: bool
    parent_aligned: bool
    pack_ready: bool
    live_dispatched: bool
    pack_dispatched: bool
    merge_executed: bool
    live_execution_authorized: bool
    charge_executed: bool
    draft_written: bool
    analysis_written: bool
    twin_written: bool
    prompts_injected: bool
    record_persisted: bool
    remote_index_queried: bool
    production_router_verdict: str
    live_router_authorized: bool
    secrets_stored: bool
    live_meter_read: bool
    purchase_executed: bool
    hosted: bool
    store_mutated: bool
    backlog_mutated: bool
    suite_rewritten: bool
    pdf_view_authorized: bool
    pdf_primary: bool
    live_dispatch_authorized: bool
    inventory_mutated: bool
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "session_id": self.session_id,
            "parent_asset_id": self.parent_asset_id,
            "week_id": self.week_id,
            "asset_id": self.asset_id,
            "title": self.title,
            "account_id": self.account_id,
            "fullscreen": self.fullscreen.to_dict(),
            "collective_pack": self.collective_pack.to_dict(),
            "session_aligned": self.session_aligned,
            "parent_aligned": self.parent_aligned,
            "pack_ready": self.pack_ready,
            "live_dispatched": False,
            "pack_dispatched": False,
            "merge_executed": False,
            "live_execution_authorized": False,
            "charge_executed": False,
            "draft_written": False,
            "analysis_written": False,
            "twin_written": False,
            "prompts_injected": False,
            "record_persisted": False,
            "remote_index_queried": False,
            "production_router_verdict": "REJECT",
            "live_router_authorized": False,
            "secrets_stored": False,
            "live_meter_read": False,
            "purchase_executed": False,
            "hosted": False,
            "store_mutated": False,
            "backlog_mutated": False,
            "suite_rewritten": False,
            "pdf_view_authorized": False,
            "pdf_primary": False,
            "live_dispatch_authorized": False,
            "inventory_mutated": False,
            "notes": list(self.notes),
            "authority": (
                "fullscreen_open_collective_multiselect_floating_dr_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise FullscreenOpenCollectiveMultiselectFloatingDrComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_fullscreen_open_collective_multiselect_floating_dr(
    *,
    fullscreen: object,
    collective_pack: object,
    operator_ack: object,
    require_both: object | None = None,
) -> FullscreenOpenCollectiveMultiselectFloatingDrCompose:
    """Fullscreen on collective multiselect floating DR. Never dispatches."""
    if not isinstance(operator_ack, bool):
        raise FullscreenOpenCollectiveMultiselectFloatingDrComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(fullscreen, dict):
        raise FullscreenOpenCollectiveMultiselectFloatingDrComposeError(
            "fullscreen must be an object"
        )
    if not isinstance(collective_pack, dict):
        raise FullscreenOpenCollectiveMultiselectFloatingDrComposeError(
            "collective_pack must be an object"
        )

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise FullscreenOpenCollectiveMultiselectFloatingDrComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "live_dispatched=false · pack_dispatched=false · merge_executed=false",
        "live_execution_authorized=false · charge_executed=false · analysis_written=false",
        "production_router_verdict=REJECT",
    ]

    try:
        fs = compose_floating_fullscreen_open(
            session_id=fullscreen.get("session_id"),
            parent_asset_id=fullscreen.get("parent_asset_id"),
            operator_ack=operator_ack,
            existing_instance=fullscreen.get("existing_instance"),
            highlight=fullscreen.get("highlight"),
            prompt=fullscreen.get("prompt"),
            gated=fullscreen.get("gated"),
            tray_siblings=fullscreen.get("tray_siblings"),
        )
    except FloatingFullscreenOpenComposeError as e:
        raise FullscreenOpenCollectiveMultiselectFloatingDrComposeError(
            str(e)
        ) from e
    notes.extend(f"[fullscreen] {n}" for n in fs.notes)

    try:
        cp = compose_collective_multiselect_floating_dr_draft_before_merge(
            multiselect=collective_pack.get("multiselect"),
            floating_dr_pack=collective_pack.get("floating_dr_pack"),
            operator_ack=operator_ack,
            require_both=collective_pack.get("require_both"),
        )
    except CollectiveMultiselectFloatingDrDraftBeforeMergeComposeError as e:
        raise FullscreenOpenCollectiveMultiselectFloatingDrComposeError(
            str(e)
        ) from e
    notes.extend(f"[collective_pack] {n}" for n in cp.notes)

    session = _require_nonempty(fs.session_id, field="session_id")
    parent = _require_nonempty(fs.parent_asset_id, field="parent_asset_id")
    week = _require_nonempty(cp.week_id, field="week_id")
    asset = _require_nonempty(cp.asset_id, field="asset_id")
    title = _require_nonempty(cp.title, field="title")
    account = _require_nonempty(cp.account_id, field="account_id")

    session_aligned = cp.session_id == session
    parent_aligned = cp.parent_asset_id == parent or cp.asset_id == parent
    if not session_aligned:
        notes.append(
            "session_id mismatch between fullscreen and collective_pack — "
            "pack_ready blocked"
        )
    else:
        notes.append("session_aligned=true")
    if not parent_aligned:
        notes.append(
            "parent_asset_id mismatch between fullscreen and collective_pack — "
            "pack_ready blocked"
        )
    else:
        notes.append("parent_aligned=true")

    if require:
        pack_ready = (
            session_aligned is True
            and parent_aligned is True
            and fs.fullscreen_ready is True
            and cp.pack_ready is True
            and cp.production_router_verdict == "REJECT"
            and fs.live_dispatched is False
            and fs.merge_executed is False
            and fs.pack_dispatched is False
            and cp.live_dispatched is False
            and cp.pack_dispatched is False
            and cp.merge_executed is False
            and cp.analysis_written is False
            and cp.twin_written is False
            and cp.draft_written is False
            and cp.live_execution_authorized is False
            and cp.charge_executed is False
            and cp.remote_index_queried is False
            and cp.pdf_primary is False
            and cp.live_router_authorized is False
            and cp.secrets_stored is False
            and operator_ack is True
        )
    else:
        pack_ready = (
            session_aligned is True
            and parent_aligned is True
            and operator_ack is True
            and cp.production_router_verdict == "REJECT"
            and cp.pdf_primary is False
            and fs.live_dispatched is False
            and (fs.fullscreen_ready is True or cp.pack_ready is True)
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — fullscreen + collective multiselect floating DR "
            "ready; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — fullscreen, collective_pack, alignment, or "
            "operator_ack gate open"
        )

    if (
        fs.live_dispatched is not False
        or fs.merge_executed is not False
        or fs.pack_dispatched is not False
        or cp.live_dispatched is not False
        or cp.pack_dispatched is not False
        or cp.merge_executed is not False
        or cp.analysis_written is not False
        or cp.twin_written is not False
        or cp.draft_written is not False
        or cp.live_execution_authorized is not False
        or cp.charge_executed is not False
        or cp.remote_index_queried is not False
        or cp.pdf_primary is not False
        or cp.live_router_authorized is not False
        or cp.secrets_stored is not False
        or cp.production_router_verdict != "REJECT"
    ):
        raise FullscreenOpenCollectiveMultiselectFloatingDrComposeError(
            "invariant: honesty flags must remain false / REJECT"
        )

    notes.extend(
        (
            "live_dispatched=false",
            "pack_dispatched=false",
            "merge_executed=false",
            "live_execution_authorized=false",
            "charge_executed=false",
            "draft_written=false",
            "analysis_written=false",
            "twin_written=false",
            "prompts_injected=false",
            "record_persisted=false",
            "remote_index_queried=false",
            "production_router_verdict=REJECT",
            "live_router_authorized=false",
            "secrets_stored=false",
            "live_meter_read=false",
            "purchase_executed=false",
            "hosted=false",
            "store_mutated=false",
            "backlog_mutated=false",
            "suite_rewritten=false",
            "pdf_view_authorized=false",
            "pdf_primary=false",
            "live_dispatch_authorized=false",
            "inventory_mutated=false",
        )
    )

    return FullscreenOpenCollectiveMultiselectFloatingDrCompose(
        session_id=session,
        parent_asset_id=parent,
        week_id=week,
        asset_id=asset,
        title=title,
        account_id=account,
        fullscreen=fs,
        collective_pack=cp,
        session_aligned=session_aligned,
        parent_aligned=parent_aligned,
        pack_ready=pack_ready,
        live_dispatched=False,
        pack_dispatched=False,
        merge_executed=False,
        live_execution_authorized=False,
        charge_executed=False,
        draft_written=False,
        analysis_written=False,
        twin_written=False,
        prompts_injected=False,
        record_persisted=False,
        remote_index_queried=False,
        production_router_verdict="REJECT",
        live_router_authorized=False,
        secrets_stored=False,
        live_meter_read=False,
        purchase_executed=False,
        hosted=False,
        store_mutated=False,
        backlog_mutated=False,
        suite_rewritten=False,
        pdf_view_authorized=False,
        pdf_primary=False,
        live_dispatch_authorized=False,
        inventory_mutated=False,
        notes=tuple(notes),
        authority=(
            "fullscreen_open_collective_multiselect_floating_dr_compose_advisory"
        ),
    )


def format_fullscreen_open_collective_multiselect_floating_dr_summary(
    c: FullscreenOpenCollectiveMultiselectFloatingDrCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"fullscreen_ready={c.fullscreen.fullscreen_ready} · "
        f"collective_ready={c.collective_pack.pack_ready} · "
        f"session_aligned={c.session_aligned} · "
        f"parent_aligned={c.parent_aligned} · "
        f"verdict={c.production_router_verdict} · "
        "live_dispatched=false · charge_executed=false · "
        "live_execution_authorized=false"
    )


__all__ = [
    "FullscreenOpenCollectiveMultiselectFloatingDrCompose",
    "FullscreenOpenCollectiveMultiselectFloatingDrComposeError",
    "compose_fullscreen_open_collective_multiselect_floating_dr",
    "format_fullscreen_open_collective_multiselect_floating_dr_summary",
]

File: apps/reading/src/api/fullscreenOpenCollectiveMultiselectFloatingDrCompose.ts
/**
 * Floating fullscreen-open residual over collective multiselect + floating DR
 * + draft-before-merge + MO price-ceiling pack (pure).
 *
 * Operator vision: open a floating deep-research instance fullscreen while
 * multi-select cohesive unit + floating DR + draft-before-merge + Midnight Oil
 * price-ceiling honesty remain pure — never live-dispatches research or MO.
 *
 * live_dispatched / pack_dispatched / merge_executed always false.
 * live_execution_authorized / charge_executed / analysis_written always false.
 * production_router_verdict always REJECT.
 */

import {
  composeFloatingFullscreenOpen,
  type FloatingFullscreenOpenCompose,
  type FloatingFullscreenOpenInput,
} from "./floatingFullscreenOpenCompose";
import {
  composeCollectiveMultiselectFloatingDrDraftBeforeMerge,
  type CollectiveMultiselectFloatingDrDraftBeforeMergeCompose,
  type CollectiveMultiselectFloatingDrDraftBeforeMergeInput,
} from "./collectiveMultiselectFloatingDrDraftBeforeMergeCompose";

export interface FullscreenOpenCollectiveMultiselectFloatingDrInput {
  fullscreen: Omit<FloatingFullscreenOpenInput, "operator_ack">;
  collective_pack: Omit<
    CollectiveMultiselectFloatingDrDraftBeforeMergeInput,
    "operator_ack"
  >;
  operator_ack: boolean;
  require_both?: boolean;
}

export interface FullscreenOpenCollectiveMultiselectFloatingDrCompose {
  session_id: string;
  parent_asset_id: string;
  week_id: string;
  asset_id: string;
  title: string;
  account_id: string;
  fullscreen: FloatingFullscreenOpenCompose;
  collective_pack: CollectiveMultiselectFloatingDrDraftBeforeMergeCompose;
  session_aligned: boolean;
  parent_aligned: boolean;
  pack_ready: boolean;
  live_dispatched: false;
  pack_dispatched: false;
  merge_executed: false;
  live_execution_authorized: false;
  charge_executed: false;
  draft_written: false;
  analysis_written: false;
  twin_written: false;
  prompts_injected: false;
  record_persisted: false;
  remote_index_queried: false;
  production_router_verdict: "REJECT";
  live_router_authorized: false;
  secrets_stored: false;
  live_meter_read: false;
  purchase_executed: false;
  hosted: false;
  store_mutated: false;
  backlog_mutated: false;
  suite_rewritten: false;
  pdf_view_authorized: false;
  pdf_primary: false;
  live_dispatch_authorized: false;
  inventory_mutated: false;
  notes: string[];
  authority: "fullscreen_open_collective_multiselect_floating_dr_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * Fullscreen-open stacked on collective multiselect floating DR pack.
 * Never live-dispatches research or MO workers.
 */
export function composeFullscreenOpenCollectiveMultiselectFloatingDr(
  input: FullscreenOpenCollectiveMultiselectFloatingDrInput,
): FullscreenOpenCollectiveMultiselectFloatingDrCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (!input.fullscreen || typeof input.fullscreen !== "object") {
    throw new Error("fullscreen must be an object");
  }
  if (!input.collective_pack || typeof input.collective_pack !== "object") {
    throw new Error("collective_pack must be an object");
  }

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "live_dispatched=false · pack_dispatched=false · merge_executed=false",
    "live_execution_authorized=false · charge_executed=false · analysis_written=false",
    "production_router_verdict=REJECT",
  ];

  const fullscreen = composeFloatingFullscreenOpen({
    ...input.fullscreen,
    operator_ack: input.operator_ack,
  });
  notes.push(...fullscreen.notes.map((n) => `[fullscreen] ${n}`));

  const collective_pack = composeCollectiveMultiselectFloatingDrDraftBeforeMerge({
    ...input.collective_pack,
    operator_ack: input.operator_ack,
  });
  notes.push(...collective_pack.notes.map((n) => `[collective_pack] ${n}`));

  const session_id = requireNonEmpty(fullscreen.session_id, "session_id");
  const parent_asset_id = requireNonEmpty(
    fullscreen.parent_asset_id,
    "parent_asset_id",
  );
  const week_id = requireNonEmpty(collective_pack.week_id, "week_id");
  const asset_id = requireNonEmpty(collective_pack.asset_id, "asset_id");
  const title = requireNonEmpty(collective_pack.title, "title");
  const account_id = requireNonEmpty(collective_pack.account_id, "account_id");

  const session_aligned = collective_pack.session_id === session_id;
  const parent_aligned =
    collective_pack.parent_asset_id === parent_asset_id ||
    collective_pack.asset_id === parent_asset_id;
  if (!session_aligned) {
    notes.push(
      "session_id mismatch between fullscreen and collective_pack — pack_ready blocked",
    );
  } else {
    notes.push("session_aligned=true");
  }
  if (!parent_aligned) {
    notes.push(
      "parent_asset_id mismatch between fullscreen and collective_pack — pack_ready blocked",
    );
  } else {
    notes.push("parent_aligned=true");
  }

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      fullscreen.fullscreen_ready === true &&
      collective_pack.pack_ready === true &&
      collective_pack.production_router_verdict === "REJECT" &&
      fullscreen.live_dispatched === false &&
      fullscreen.merge_executed === false &&
      fullscreen.pack_dispatched === false &&
      collective_pack.live_dispatched === false &&
      collective_pack.pack_dispatched === false &&
      collective_pack.merge_executed === false &&
      collective_pack.analysis_written === false &&
      collective_pack.twin_written === false &&
      collective_pack.draft_written === false &&
      collective_pack.live_execution_authorized === false &&
      collective_pack.charge_executed === false &&
      collective_pack.remote_index_queried === false &&
      collective_pack.pdf_primary === false &&
      collective_pack.live_router_authorized === false &&
      collective_pack.secrets_stored === false &&
      input.operator_ack === true;
  } else {
    pack_ready =
      session_aligned &&
      parent_aligned &&
      input.operator_ack === true &&
      collective_pack.production_router_verdict === "REJECT" &&
      collective_pack.pdf_primary === false &&
      fullscreen.live_dispatched === false &&
      (fullscreen.fullscreen_ready === true ||
        collective_pack.pack_ready === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — fullscreen + collective multiselect floating DR ready; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — fullscreen, collective_pack, alignment, or operator_ack gate open",
    );
  }

  if (
    fullscreen.live_dispatched !== false ||
    fullscreen.merge_executed !== false ||
    fullscreen.pack_dispatched !== false ||
    collective_pack.live_dispatched !== false ||
    collective_pack.pack_dispatched !== false ||
    collective_pack.merge_executed !== false ||
    collective_pack.analysis_written !== false ||
    collective_pack.twin_written !== false ||
    collective_pack.draft_written !== false ||
    collective_pack.live_execution_authorized !== false ||
    collective_pack.charge_executed !== false ||
    collective_pack.remote_index_queried !== false ||
    collective_pack.pdf_primary !== false ||
    collective_pack.live_router_authorized !== false ||
    collective_pack.secrets_stored !== false ||
    collective_pack.production_router_verdict !== "REJECT"
  ) {
    throw new Error("invariant: honesty flags must remain false / REJECT");
  }

  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");
  notes.push("merge_executed=false");
  notes.push("live_execution_authorized=false");
  notes.push("charge_executed=false");
  notes.push("draft_written=false");
  notes.push("analysis_written=false");
  notes.push("twin_written=false");
  notes.push("prompts_injected=false");
  notes.push("record_persisted=false");
  notes.push("remote_index_queried=false");
  notes.push("production_router_verdict=REJECT");
  notes.push("live_router_authorized=false");
  notes.push("secrets_stored=false");
  notes.push("live_meter_read=false");
  notes.push("purchase_executed=false");
  notes.push("hosted=false");
  notes.push("store_mutated=false");
  notes.push("backlog_mutated=false");
  notes.push("suite_rewritten=false");
  notes.push("pdf_view_authorized=false");
  notes.push("pdf_primary=false");
  notes.push("live_dispatch_authorized=false");
  notes.push("inventory_mutated=false");

  return {
    session_id,
    parent_asset_id,
    week_id,
    asset_id,
    title,
    account_id,
    fullscreen,
    collective_pack,
    session_aligned,
    parent_aligned,
    pack_ready,
    live_dispatched: false,
    pack_dispatched: false,
    merge_executed: false,
    live_execution_authorized: false,
    charge_executed: false,
    draft_written: false,
    analysis_written: false,
    twin_written: false,
    prompts_injected: false,
    record_persisted: false,
    remote_index_queried: false,
    production_router_verdict: "REJECT",
    live_router_authorized: false,
    secrets_stored: false,
    live_meter_read: false,
    purchase_executed: false,
    hosted: false,
    store_mutated: false,
    backlog_mutated: false,
    suite_rewritten: false,
    pdf_view_authorized: false,
    pdf_primary: false,
    live_dispatch_authorized: false,
    inventory_mutated: false,
    notes,
    authority:
      "fullscreen_open_collective_multiselect_floating_dr_compose_advisory",
  };
}

export function formatFullscreenOpenCollectiveMultiselectFloatingDrSummary(
  c: FullscreenOpenCollectiveMultiselectFloatingDrCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `fullscreen_ready=${c.fullscreen.fullscreen_ready} · ` +
    `collective_ready=${c.collective_pack.pack_ready} · ` +
    `session_aligned=${c.session_aligned} · ` +
    `parent_aligned=${c.parent_aligned} · ` +
    `verdict=${c.production_router_verdict} · ` +
    `live_dispatched=false · charge_executed=false · live_execution_authorized=false`
  );
}
